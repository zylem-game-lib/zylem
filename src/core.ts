export { game } from './lib/game/game';
export type { ZylemGameConfig } from './lib/game/game-interfaces';

export { vessel } from './lib/core/vessel';

export type { Vect3 } from './lib/core/utility';

export { globalChange, globalChanges, variableChange, variableChanges } from './lib/actions/global-change';

