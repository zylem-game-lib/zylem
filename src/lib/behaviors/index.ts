// export * from './components/bounce';
// export * from './components/move';
// export * from './components/transform';
// export * from './test-system';

export * from '../actions/update/ricochet';
export * from '../actions/collision/ping-pong';