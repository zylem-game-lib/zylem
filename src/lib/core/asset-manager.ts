// Use service worker when loading assets