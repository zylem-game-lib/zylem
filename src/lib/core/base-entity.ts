import {
	SetupFunction,
	UpdateFunction,
	DestroyFunction,
	BaseEntityOptions,
} from '../interfaces/entity';
import { Entity, EntityParameters } from './entity';
import { nanoid } from 'nanoid';

export class BaseEntity<T> implements Entity<T> {
	public uuid: string;
	public _custom: { [key: string]: any };

	protected type: string = 'BaseEntity';
	protected _setup: SetupFunction<T>;
	protected _update: UpdateFunction<T>;
	protected _destroy: DestroyFunction<T>;

	static entityCount = 1;

	constructor(options: BaseEntityOptions<T>) {
		this.uuid = `${++BaseEntity.entityCount}-${nanoid()}`;
		this._setup = options.setup || (() => { });
		this._update = options.update || (() => { });
		this._destroy = options.destroy || (() => { });
		this._custom = options.custom || {};
	}

	public create(): Promise<T> {
		throw new Error('Create method needs to be implemented for Entities.');
	}

	public setup(_: EntityParameters<T>): void { }

	public update(_: EntityParameters<T>): void { }

	public destroy(_: EntityParameters<T>): void { }

}
