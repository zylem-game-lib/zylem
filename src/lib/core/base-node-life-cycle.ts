import { ZylemHUD } from '../ui/hud';
import { ZylemCamera } from '../camera/camera';
import { Game } from './game-wrapper';
import { Inputs } from '../input/input';

export interface SetupContext<T> {
	entity: T;
	globals: any;
	inputs?: Inputs;
	HUD?: ZylemHUD;
	camera?: ZylemCamera;
	game?: Game;
}

export interface SetupFunction<T> {
	(context: SetupContext<T>): void;
}

export type UpdateContext<T> = {
	entity: T;
	delta: number;
	inputs: Inputs;
	globals: any;
	camera: ZylemCamera;
	game?: Game;
	HUD?: ZylemHUD;
};

export interface UpdateFunction<T> {
	(context: UpdateContext<T>): void;
}

export interface DestroyContext<T> {
	entity: T;
	globals: any;
}

export interface DestroyFunction<T> {
	(context: DestroyContext<T>): void;
}