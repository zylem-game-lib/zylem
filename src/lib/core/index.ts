// Core exports only - entities moved to separate exports to prevent circular dependencies
export * from '../game/zylem-game';
export * from '../stage/zylem-stage';
export * from './vessel';

// Export shared types
export * from '../types';