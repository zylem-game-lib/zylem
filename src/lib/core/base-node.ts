import { Behavior } from "~/lib/actions/behaviors/behavior";
import {
	CleanupContext,
	CleanupFunction,
	DestroyContext,
	DestroyFunction,
	LoadedContext,
	LoadedFunction,
	SetupContext,
	SetupFunction,
	UpdateContext,
	UpdateFunction,
} from "./base-node-life-cycle";
import { DEBUG_FLAG } from "./flags";
import { nanoid } from "nanoid";

export type BaseNodeOptions<T = any> = BaseNode | Partial<T>;

export abstract class BaseNode<Options = any, T = any> {
	protected parent: BaseNode | null = null;
	protected children: BaseNode[] = [];
	public behaviors: Behavior[] = [];
	public options: Options;
	public eid: number = 0;
	public uuid: string = '';
	public name: string = '';
	public markedForRemoval: boolean = false;

	setup: SetupFunction<this> = () => { };
	loaded: LoadedFunction<this> = () => { };
	update: UpdateFunction<this> = () => { };
	destroy: DestroyFunction<this> = () => { };
	cleanup: CleanupFunction<this> = () => { };

	constructor(args: BaseNodeOptions[] = []) {
		const options = args
			.filter(arg => !(arg instanceof BaseNode))
			.reduce((acc, opt) => ({ ...acc, ...opt }), {});
		this.options = options as Options;
		this.uuid = nanoid();
	}

	public setParent(parent: BaseNode | null): void {
		this.parent = parent;
	}

	public getParent(): BaseNode | null {
		return this.parent;
	}

	public add(baseNode: BaseNode): void {
		this.children.push(baseNode);
		baseNode.setParent(this);
	}

	public remove(baseNode: BaseNode): void {
		const index = this.children.indexOf(baseNode);
		if (index !== -1) {
			this.children.splice(index, 1);
			baseNode.setParent(null);
		}
	}

	public getChildren(): BaseNode[] {
		return this.children;
	}

	public isComposite(): boolean {
		return this.children.length > 0;
	}

	public abstract create(): T;

	protected abstract _setup(params: SetupContext<this>): void;

	protected abstract _loaded(params: LoadedContext<this>): Promise<void>;

	protected abstract _update(params: UpdateContext<this>): void;

	protected abstract _destroy(params: DestroyContext<this>): void;

	protected abstract _cleanup(params: CleanupContext<this>): Promise<void>;

	public nodeSetup(params: SetupContext<this>) {
		if (DEBUG_FLAG) { /**  */ }
		if (typeof this._setup === 'function') {
			this._setup(params);
		}
		if (this.setup) {
			this.setup(params);
		}
		this.children.forEach(child => child.nodeSetup(params));
	}

	public nodeUpdate(params: UpdateContext<this>): void {
		if (this.markedForRemoval) {
			return;
		}
		if (typeof this._update === 'function') {
			this._update(params);
		}
		if (this.update) {
			this.update(params);
		}
		this.children.forEach(child => child.nodeUpdate(params));
	}

	public nodeDestroy(params: DestroyContext<this>): void {
		this.children.forEach(child => child.nodeDestroy(params));
		if (this.destroy) {
			this.destroy(params);
		}
		if (typeof this._destroy === 'function') {
			this._destroy(params);
		}
		this.markedForRemoval = true;
	}

	public getOptions(): Options {
		return this.options;
	}

	public setOptions(options: Partial<Options>): void {
		this.options = { ...this.options, ...options };
	}
}