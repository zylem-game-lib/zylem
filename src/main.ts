import * as THREE from 'three';
import * as RAPIER from '@dimforge/rapier3d-compat';
import { Howl } from 'howler';

import { game, IGameOptions, stage, StageOptions } from './lib/core';

import { PerspectiveType, Perspectives } from './lib/interfaces/perspective';
import { Vect3 } from './lib/core/utility';

import * as actions from './lib/behaviors/actions';
import * as entities from './lib/entities/index';

const Util = {
	...actions
};

const Zylem = {
	Util
};

const { box, sphere, sprite, plane, zone, actor, vessel, ZylemBox } = entities;

namespace Zylem { };

export {
	game,
	stage,
	box,
	sphere,
	sprite,
	plane,
	zone,
	actor,
	vessel,
	actions,
	Perspectives,
	Zylem,
	ZylemBox,
	Howl,
	THREE,
	RAPIER
};
export type { IGameOptions as ZylemGame, StageOptions as ZylemStage, Vect3, PerspectiveType };
