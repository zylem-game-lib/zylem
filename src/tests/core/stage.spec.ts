import { describe, expect, it } from 'vitest';
import { stage } from '../../main';
import { Color } from 'three';
import { ZylemStageConfig } from '~/lib/core';

describe('create a basic stage', () => {
	it('default configuration', () => {
		const result = stage();
		result.stageRef = null as any;
		expect(result).toMatchSnapshot();
	});
	it('should create a stage with a blue background', async () => {
		const result = stage({ backgroundColor: new Color(Color.NAMES.blue) });
		const options = result.options[0] as ZylemStageConfig;
		expect(options.backgroundColor).toEqual(new Color(Color.NAMES.blue));
	});
});
