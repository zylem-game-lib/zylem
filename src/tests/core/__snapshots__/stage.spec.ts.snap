// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create a basic stage > default configuration 1`] = `
Stage {
  "destroy": [Function],
  "options": [],
  "setup": [Function],
  "stageRef": null,
  "update": [Function],
}
`;
